dartagnan/src/main/java/com/dat3m/dartagnan/solver/caat4wmm/ExecutionGraph.java

```diff
@@ -198,9 +222,19 @@ private RelationGraph getOrCreateGraphFromRelation(Relation rel) {
         if (relationGraphMap.containsKey(rel)) {
             return relationGraphMap.get(rel);
         }
+        RelationGraph graph = createGraphFromRelation(rel);
+        relationGraphMap.put(rel, graph);
+        return graph;
+    }
 
+    private RelationGraph createGraphFromRelation(Relation rel) {
         RelationGraph graph;
-        Class<?> relClass = rel.getClass();
+        Class<?> relClass = rel.getDefinition().getClass();
+        List<Relation> dependencies = rel.getDependencies();
+        RelationGraph[] graphs = new RelationGraph[dependencies.size()];
+        for (int i = 0; i < graphs.length; i++) {
+            graphs[i] = getOrCreateGraphFromRelation(dependencies.get(i));
+        }
```

Comment:
The early construction of the dependencies does not interact well with cutting and the check for special relations, because now dependencies get needlessly constructed. Fortunately, this causes only computation overhead but does not affect the result of CAAT in any way.

Reply:
I moved dependency transformation after the special and cut-relations tests.